Improve error handling for Powerwall

jrester opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This improves error handling and reporting for when the Powerwall is not available or uses an unknown software version.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jrester committed:
Improve error handling
	modified:   homeassistant/components/powerwall/__init__.py
	modified:   homeassistant/components/powerwall/config_flow.py
	modified:   homeassistant/components/powerwall/const.py
	modified:   homeassistant/components/powerwall/strings.json
	modified:   homeassistant/components/powerwall/translations/en.json

jrester committed:
Change exception name
	modified:   homeassistant/components/powerwall/__init__.py
	modified:   homeassistant/components/powerwall/config_flow.py

jrester committed:
Add test

bdraco commented:
/AzurePipelines run

bdraco commented:
/AzurePipelines run

bdraco commented on homeassistant/components/powerwall/__init__.py:
```diff
@@ -122,6 +156,7 @@ async def async_update_data():
             POWERWALL_OBJECT: power_wall,
             POWERWALL_COORDINATOR: coordinator,
             POWERWALL_HTTP_SESSION: http_session,
+            PowerwallError: None,
```
```suggestion
            POWERWALL_ERROR: None,
```

Looks like `PowerwallError` is an exception?

jrester replied:
Ups, seems like the IDE autocompleted that the wrong way...

bdraco reviewed (changes requested):
Looks like there is a key that needs to be updated

jrester committed:
Rename PowerwallError to POWERWALL_ERROR

jrester commented:
Yea, seems like the IDE tried to autocomplete it... :grimacing:

bdraco commented on homeassistant/components/powerwall/__init__.py:
```diff
@@ -87,26 +93,54 @@ def _async_migrator(entity_entry: entity_registry.RegistryEntry):
+    except PowerwallError as err:
         http_session.close()
+        await handle_setup_error(hass, entry, err)
         raise ConfigEntryNotReady
```
@jrester The system is going to retry forever when `ConfigEntryNotReady` is raised.  We should only raise this if we expect the situation will resolve such as a transient network issue.  

I'm thinking `PowerwallUnreachableError` and `ConnectionError` get `ConfigEntryNotReady` and `APIError` gets the new code?

jrester replied:
Inside `handle_setup_error` it is first checked if we already had an error and only if it is new the notification is sent. This is then reset after the component was successfully setup. But I can also remove the notification part. Was quite unsure if it would be useful to inform the user about such failures.

But it might be sensible to completely refuse setup when an APIChangedError is received, because this means the API cannot be queried anymore.

jrester replied:
Is there any way to terminate the `DataUpdateCoordinator` from the `async_update_data` method? Because if we receive an `APIChangedError` we should completely stop polling.

bdraco replied:
@jrester AFAICT `DataUpdateCoordinator` doesn't contemplate that scenario.  We could set a flag `hass.data[DOMAIN][entry.entry_id][UPDATE_PERM_FAILED] = True` (or check the error) and have `async_update_data` return when its set so the I/O is avoided.  Its not as clean as I'd like but is also like you said, an edge case.

jrester replied:
@bdraco The problem is that AFAIC we need to return correctly structured data otherwise the log gets filled with `TypeError`s because of the sensors trying to display it. I would return cached data but this could still fill the logs when it is the first request that is performed. The easiest workaround could be to request all data during the startup. But I am quite unsure if it is really worth the effort.

bdraco replied:
@jrester Maybe something like this? https://github.com/bdraco/home-assistant/commit/adb659ba2b7ba22915daa4d6306a80d9ccd1a276

I haven't see where the `TypeError`s are coming from so I may be missing the issue though.

jrester replied:
@bdraco I had a similar idea but when the `async_update_data` returns None this gets set as data for the `DataUpdateCoordinator` and when the senor tries to use the data for example here `self._coordinator.data[POWERWALL_API_CHARGE]` this raises an error because `self._coordinator.data` is None and as such isn't subscriptable.

bdraco replied:
If we `raise UpdateFailed("Powerwall api has changed")` instead of returning `None` does that work around the issue?

jrester committed:
Modify handling of APIChangedErrors
	modified:   homeassistant/components/powerwall/__init__.py
	modified:   homeassistant/components/powerwall/const.py

bdraco commented:
/AzurePipelines Run

bdraco reviewed (approved):
Looks good, everything still working as expected

MartinHjelmare commented on homeassistant/components/powerwall/__init__.py:
```diff
@@ -97,16 +109,37 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
         """Fetch data from API endpoint."""
-        return await hass.async_add_executor_job(_fetch_powerwall_data, power_wall)
+        # Check if we had an error before
+        _LOGGER.info("Checking if update failed")
```
This should be debug level.

MartinHjelmare commented on homeassistant/components/powerwall/__init__.py:
```diff
@@ -97,16 +109,37 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+        # Check if we had an error before
+        _LOGGER.info("Checking if update failed")
+        if not hass.data[DOMAIN][entry.entry_id][POWERWALL_API_CHANGED]:
+            _LOGGER.info("Updating data")
```
Here too.
